File: ApprovalTestKoans/ApprovalTestKoans/Lesson04/UsingReporters.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Reflection;
using ApprovalTestKoans.Helpers;
using ApprovalTests.Core;
using ApprovalTests.Reporters;
using ApprovalUtilities.Utilities;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using Approvals = ApprovalTests.Approvals;

namespace ApprovalTestKoans.Lesson04
{
    /*
     * Note: If the windows that appear keep closing, make sure to run the test with 
     * Visual Studio instead of ReSharper.
     */

    [TestClass]
	[UseReporter(typeof (MsTestReporter))]
	public class ConfiguringReporters : Koans
	{
		[TestMethod]
		[UseReporter(typeof (_____))]
		public void ConfiguringTheFileLauncherReporterAtTheMethodLevel()
		{
			var reporter = Approvals.GetReporter();
			Assert.IsInstanceOfType(reporter, typeof (FileLauncherReporter));
		}

		[TestMethod]
		public void ConfiguringAClassLevelDefault()
		{
			var reporter = Approvals.GetReporter();
			Assert.IsInstanceOfType(reporter, typeof (_____));
		}
	}

	[TestClass]
	public class ConfiguringReportersPart2 : Koans
	{
		[TestMethod]
		public void ConfiguringAnAssemblyLevelDefault()
		{
			var reporter = Approvals.GetReporter();
			Assert.IsInstanceOfType(reporter, typeof (DiffReporter));
			// Hint: The blank for this problem is in the AssemblyInfo.cs file
		}
	}

	[TestClass]
	public class UsingReporters : Koans
	{
		[TestMethod]
		[UseReporter(typeof (ClipboardReporter), typeof (ImageWebReporter), typeof (TortoiseImageDiffReporter))]
		public void ConfiguringMultipleReporters()
		{
			var reporter = Approvals.GetReporter();
			MultiReporter multi = (MultiReporter) reporter;
			var second = multi.GetReporters(1);
			Assert.IsInstanceOfType(second, typeof (_____));
		}

		[TestMethod]
		[UseReporter(typeof (FileLauncherReporter))]
		public void UsingTextReportersForInsight()
		{
			if (decode("=fikpKnf") != ___)
			{
				var reporter = Approvals.GetReporter();
				reporter.Report("", PathUtilities.GetAdjacentFile("Insight.txt"));
				Assert.Fail("Please fill in the blank");
			}
		}

		[TestMethod]
		[UseReporter(typeof (ImageWebReporter))]
		public void UsingImageForInsight()
		{
			String whatWasTheOldColor = ___;
			if (decode("Ycl\\") != whatWasTheOldColor.ToLowerInvariant())
			{
				var reporter = Approvals.GetReporter();
				reporter.Report(PathUtilities.GetAdjacentFile("OldImage.png"),PathUtilities.GetAdjacentFile("NewImage.png"));
				Assert.Fail("Please fill in the blank");
			}
		}
	}

	public class _____ : IApprovalFailureReporter
	{
		public void Report(string approved, string received)
		{
			throw new NotImplementedException();
		}
	}

	public class ImageWebReporter : IEnvironmentAwareReporter
	{
		public static ImageWebReporter INSTANCE = new ImageWebReporter();

		public void Report(string approved, string received)
		{
			String text =
				"<html><body><center><table border=1><tr><td><img src=\"file:///{0}\"></td><td><img src=\"file:///{1}\"></td></tr><tr><td>approved</td><td>received</td></tr></table>  {2} <br/> <b>to approve :</b> copy clipboard to command window  <br/> <font size=1>{3}</font></center></body></html>";
			String moveText = QuietReporter.GetCommandLineForApproval(approved, received);
			text = text.FormatWith(approved, received, received, moveText);
			string file = Path.GetTempFileName() + ".html";
			File.WriteAllText(file, text);
			Process.Start(file);
		}

		public bool IsWorkingInThisEnvironment(string forFile)
		{
			return GenericDiffReporter.IsFileOneOf(forFile, GenericDiffReporter.GetImageFileTypes());
		}
	}

	public static class ReporterHelper
	{
		public static IApprovalFailureReporter GetReporters(this MultiReporter reporter, int index)
		{
			FieldInfo field = typeof (MultiReporter).GetField("reporters", BindingFlags.NonPublic | BindingFlags.Instance);
			var value = (IEnumerable<IApprovalFailureReporter>) field.GetValue(reporter);
			return value.ToArray()[index];
		}
	}
}

File: ApprovalTests.Net.Koans/ApprovalTests.Net.Koans/Lesson04/UsingReporters.java
package org.approvaltests.koans.lesson04;

import org.approvaltests.ReporterFactory;
import org.approvaltests.core.ApprovalFailureReporter;
import org.approvaltests.koans.helpers.Koans;
import org.approvaltests.koans.helpers._____;
import org.approvaltests.reporters.ClipboardReporter;
import org.approvaltests.reporters.FileLauncherReporter;
import org.approvaltests.reporters.ImageWebReporter;
import org.approvaltests.reporters.MultiReporter;
import org.approvaltests.reporters.QuietReporter;
import org.approvaltests.reporters.UseReporter;
import org.approvaltests.reporters.windows.TortoiseImageDiffReporter;
import org.junit.Assert;
import org.junit.Test;

/**
 * How to do Koans:
 * Step 1: Press the Run Button (Place cursor on the Method name to run a single method)
 * PC: Ctrl+F11
 * Mac: Command+fn+F11
 * Step 2: Read the name of the Method that Failed
 * Step 3: Fill in the blank (___) to make it pass
 * Step 4: Repeat Until Enlightenment
 * Do not change anything except the blank (___)
 */
@UseReporter(QuietReporter.class)
public class UsingReporters extends Koans
{
  @Test
  @UseReporter(_____.class)
  public void ConfiguringTheFileLauncherReporter() throws Exception
  {
    ApprovalFailureReporter reporter = ReporterFactory.getFromAnnotation();
    Assert.assertTrue("Please fill in the blank", reporter instanceof FileLauncherReporter);
  }
  @Test
  public void ConfiguringAClassLevelDefault() throws Exception
  {
    ApprovalFailureReporter reporter = ReporterFactory.getFromAnnotation();
    Assert.assertTrue(reporter.getClass().getName(), reporter instanceof _____);
  }
  @Test
  @UseReporter({ClipboardReporter.class, ImageWebReporter.class, TortoiseImageDiffReporter.class})
  public void ConfiguringMultipleReporters() throws Exception
  {
    ApprovalFailureReporter reporter = ReporterFactory.getFromAnnotation();
    MultiReporter multi = (MultiReporter) reporter;
    ApprovalFailureReporter second = multi.getReporters()[1];
    Assert.assertTrue(second.getClass().getName(), second instanceof _____);
  }
  @Test
  @UseReporter(FileLauncherReporter.class)
  public void UsingTextReportersForInsight() throws Exception
  {
    if (!decode("=fikpKnf").equals(____))
    {
      ApprovalFailureReporter reporter = ReporterFactory.getFromAnnotation();
      reporter.report(getPath("Insight.txt"), "");
      Assert.fail("Please fill in the blank");
    }
  }
  @Test
  @UseReporter(ImageWebReporter.class)
  public void UsingImageForInsight() throws Exception
  {
    String whatWasTheOldColor = ___;
    if (!decode("Ycl\\").equals(whatWasTheOldColor.toLowerCase()))
    {
      ApprovalFailureReporter reporter = ReporterFactory.getFromAnnotation();
      reporter.report(getPath("NewImage.png"), getPath("OldImage.png"));
      Assert.fail("Please fill in the blank");
    }
  }
}
